### PR TITLE
Lifecycle and navigation testing modules

### PR DESCRIPTION
**Pull request by terrakok**

### CODE REVIEW

File: lifecycle/lifecycle-runtime-testing/src/commonMain/kotlin/androidx/lifecycle/testing/TestLifecycleOwner.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2024 The Android Open Source Project
```
Comment (MatkovIvan) [inline lifecycle/lifecycle-runtime-testing/src/commonMain/kotlin/androidx/lifecycle/testing/TestLifecycleOwner.kt:2]: Why is it not moved as-is from Android? Is it a copy from AOSP KMP adoption?

File: navigation/navigation-common/src/commonMain/kotlin/androidx/navigation/NoOpNavigator.kt [hunk 2]
```diff
@@ -20,7 +20,7 @@ import androidx.core.bundle.Bundle
 
 /** A [Navigator] that only supports creating destinations. */
 @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
-public expect class NoOpNavigator : Navigator<NavDestination> {
+public expect class NoOpNavigator() : Navigator<NavDestination> {
```
Comment (MatkovIvan) [inline navigation/navigation-common/src/commonMain/kotlin/androidx/navigation/NoOpNavigator.kt:23]: `actual constructor` too

File: testutils/testutils-navigation/src/commonMain/kotlin/androidx/testutils/TestNavigatorDestinationBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+inline fun NavGraphBuilder.test(route: String) = test(route) {}
+
+/**
+ * Construct a new [TestNavigator.Destination]
+ */
+inline fun NavGraphBuilder.test(
+    route: String,
+    builder: TestNavigatorDestinationBuilder.() -> Unit
+) = destination(
+    TestNavigatorDestinationBuilder(provider["test"/* TEST_NAVIGATOR_NAME */], route).apply(builder)
```
Comment (MatkovIvan) [inline testutils/testutils-navigation/src/commonMain/kotlin/androidx/testutils/TestNavigatorDestinationBuilder.kt:41]: It seems that `TEST_NAVIGATOR_NAME` should be available here

File: lifecycle/lifecycle-runtime-testing/build.gradle [hunk 4]
```diff
@@ -21,38 +21,101 @@
...
+        jbTest.dependsOn(commonTest)
+        desktopTest.dependsOn(jvmTest)
+        desktopTest.dependsOn(jbTest)
+        nativeTest.dependsOn(jbTest)
+        webTest.dependsOn(jbTest)
+        jsTest.dependsOn(webTest)
+        wasmJsTest.dependsOn(webTest)
 
+        targets.all { target ->
+            if (target.platformType == org.jetbrains.kotlin.gradle.plugin.KotlinPlatformType.native) {
```
Comment (MatkovIvan) [inline lifecycle/lifecycle-runtime-testing/build.gradle]:
```kt
import org.jetbrains.kotlin.gradle.plugin.KotlinPlatformType
```
and

```suggestion
            if (target.platformType == KotlinPlatformType.native) {
```

File: lifecycle/lifecycle-runtime-testing/bcv/native/current.txt [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Klib ABI Dump
```
Comment (MatkovIvan) [inline lifecycle/lifecycle-runtime-testing/bcv/native/current.txt:1]: Do we already have native API validation? Or is it a copy from AOSP?
cc @igordmn

File: lifecycle/lifecycle-runtime-testing/src/commonMain/kotlin/androidx/lifecycle/testing/TestLifecycleOwner.kt (hunk 1 above)
Comment (terrakok) [reply lifecycle/lifecycle-runtime-testing/src/commonMain/kotlin/androidx/lifecycle/testing/TestLifecycleOwner.kt:2]: yes. I didn't change aosp commit

File: lifecycle/lifecycle-runtime-testing/bcv/native/current.txt (hunk 5 above)
Comment (terrakok) [reply lifecycle/lifecycle-runtime-testing/bcv/native/current.txt:1]: it is from the cherry-pick

File: testutils/testutils-navigation/src/commonMain/kotlin/androidx/testutils/TestNavigatorDestinationBuilder.kt (hunk 3 above)
Comment (terrakok) [reply testutils/testutils-navigation/src/commonMain/kotlin/androidx/testutils/TestNavigatorDestinationBuilder.kt:41]: No. This is a public inline function, but `TEST_NAVIGATOR_NAME` is an internal constant. it is not possible to use it here.

File: navigation/navigation-common/src/commonMain/kotlin/androidx/navigation/NoOpNavigator.kt (hunk 2 above)
Comment (terrakok) [reply navigation/navigation-common/src/commonMain/kotlin/androidx/navigation/NoOpNavigator.kt:23]: I don't get it. It is an expect declaration. Every actual constructor is marked as actual of course

Comment (MatkovIvan) [reply navigation/navigation-common/src/commonMain/kotlin/androidx/navigation/NoOpNavigator.kt:23]: sorry, nvm, mixed-up things - I thought that android's `actual` 😵‍💫

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**